### ISSUE TITLE
aria-label field on insert/edit link

### ISSUE THREAD

**Issue by lorig** [tinymce/tinymce#5037]
Feature request:

Are there any plans to include a field on the insert/edit link interface so one could include an aria-label? We are required to comply with WCAG standards and this would help us immensely. For example:

`<a href="someURL.edu" aria-label="MBA admissions">Admissions</a>`

![tiny-mce-aria-label](https://user-images.githubusercontent.com/5813187/59270710-30f95700-8c20-11e9-93f7-fda4e1313a7c.jpg)

We have both the stand-alone plugin for TinyMCE Advanced (5.2) and also WPBakery with TinyMCE integrated on different domains. Our WordPress (5.2.1) version and all plugins are all up to date. Looks like the majorVersion is 4 and the minorVersion is 9.4 for both.

**Comment by lnewson** [comment]
@lorig could you please add some more details why you'd want this and can't just use the title field, or update the text to match? As it's generally a bad idea to have the accessible text be something different to what you can see, as the experience should be kept the same (or as close as possible) regardless of if you are using a screen reader or just reading directly.

**Comment by TheSpyder** [comment]
Also the title field has other advantages such as providing a tooltip not just text for screen readers.

That screenshot looks like it is the custom Wordpress link dialog, not a TinyMCE dialog. I think we might have to close this and recommend logging it with Wordpress directly.

**Comment by lorig** [comment]
@lnewson and @TheSpyder. Thank you for your replies. While you are right that using link text which describes where the link is taking you is the best way to be accessible, there are cases where you might have multiple links on a page that have the same text, but go to different places. Including descriptive text for all of these links becomes verbose, so using an aria-label is the solution because screen readers will read that text instead of the link text. It is my understanding that the title attribute is no subsitute for an aria-label and not all screen readers will read title text, even though it offers a hover-over tooltip.

I apologize for logging this in TinyMCE, as I was not aware that it was a WordPress interface since I am using the TinyMCE editor to insert the links. Thank you for letting me know.

**Comment by m-bosch** [comment]
The aria-label SHOULD be added to the editor when you put a link arround an image for example.

See: https://www.w3.org/WAI/GL/wiki/Using_aria-label_for_link_purpose#Examples.

For WCAG2.1 targets it would be nice that feature that @lorig suggests will be implemented.

If this is already decided, is there a roadmap when?

**Comment by afercia** [comment]
> The aria-label SHOULD be added to the editor when you put a link around an image for example.

Not sure 😄 In that case, the image `alt` attribute is actually the link content and should be used to convey the link purpose or destination.

Generally, ARIA should be avoided when there's the option to use plain text for the same purpose.

Also, this would put great power and great responsibility in the hands of content creators. For example, an aria-label that is completely different from the visible text would make impossible to activate the link for speech recognition software users.

Reference:
Label in Name 
https://www.w3.org/TR/WCAG21/#label-in-name

**Comment by afercia** [comment]
Noting that the example in the original issue description:

`<a href="someURL.edu" aria-label="MBA admissions">Admissions</a>`

Would _not_ work with speech recognition software.

**Comment by m-bosch** [comment]
@afercia this should be done where there is an image in indeed. i agree this wont work well with text in the live.

So the link title should do the work...

**Comment by lorig** [comment]
> Noting that the example in the original issue description:
> 
> `<a href="someURL.edu" aria-label="MBA admissions">Admissions</a>`
> 
> Would _not_ work with speech recognition software.

An aria-label, when present, is read instead of the actual link text and can be used to differentiate links when the link text is the same for more than one instance on a page. For example, you have different sections on a page that to a non-visually impaired person, the groupings are very clear, but for a screen reader tabbing through the page, the aria-label is read instead of "Admissions...Admissions...Admissions". Link title attributes do not work.

**Comment by lorig** [comment]
I agree that using descriptive link text is best, however, there are times when that is not ideal given space limitations or verbose text content.

We use Siteimprove and they flag accessibility errors like this where "Two Year" is linked up three times on the page, but each instance goes to a different location. Here is a simple example:
Full-Time MBA
- [Two-Year](url)
- [One-Year All Majors](url)
- [One-Year Business Majors](url)
Online MBA
- [Two-Year](url)
- [Hybrid One-Year](url)
Weekend Professional MBA
- [Two-Year](url)
- [Two-Year South Florida](url)

The only way around this is to be verbose:
Full-Time MBA
- [Full-Time MBA: Two Year](url)
- [Full-Time MBA: All Majors](url)
- [Full-Time MBA: One-Year Business Majors](url)
Online MBA
- [Online MBA: Two-Year](url)
- [Online MBA: Hybrid One-Year](url)
Weekend Professional MBA
- [Weekend Professional MBA: Two-Year](url)
- [Weekend Professional MBA: Two-Year South Florida](url)

This affects our menus as well, where being verbose like the above example is out of the question for our non-visually impaired users. Being able to add an aria-label would solve this issue while allowing visually impaired users to tab through the page from link to link and know exactly where the link would be taking them. Siteimprove says the following...

> Link text used for multiple different destinations
> Description of this issue:
> The same link text is used for links going to different destinations. Users might not know the difference if they are not somehow explained.
> How to fix it:
> Make sure links are distinguishable by just their link texts or WAI-ARIA labels ('aria-labelledby' or 'aria-label') to make it clear that they lead to different destinations.

One of the methods to fix this issue: https://www.w3.org/TR/WCAG20-TECHS/H33.html

We are using some custom elements and have included a field for things like buttons so people can add an aria-label and it has worked out well.

I would imagine that anyone who doesn't know how to use the aria-label knows nothing of accessibility and would likely leave it blank. I have done my best to educate our website editors on this so they use it properly, however, ideally, there could be a tip icon that instructs people what the field is for. I am not sure how this could do more harm than good when a person is editing a website and could do more harm than good at any point in their editing. Perhaps the editing view could even show what the aria-label is so it doesn't get forgotten when the link changes.

For those of us required to make our website accessible, this would be a big help.

**Comment by afercia** [comment]
> An aria-label, when present, is read instead of the actual link text and can be used to differentiate links when the link text is the same for more than one instance on a page. For example, you have different sections on a page that to a non-visually impaired person, the groupings are very clear, but for a screen reader tabbing through the page, the aria-label is read instead of "Admissions...Admissions...Admissions". Link title attributes do not work.

Thanks @lorig, I'm pretty familiar with how aria-label works 🙂 Still, that example wouldn't work for speech recognition software users. Screen readers are just _one_ of the assistive technologies to support. 

To meet the [Label in Name](https://www.w3.org/TR/WCAG21/#label-in-name) WCAG success criterion, the aria-label needs to _at least_ start with the same words used for the visible text. Otherwise, it won't work for all users (and it will also be flagged as non-compliant). 

> I am not sure how this could do more harm than good

The example you provided at the beginning of this issue would:

`<a href="someURL.edu" aria-label="MBA admissions">Admissions</a>`

**Comment by lorig** [comment]
@afercia You are correct and I was mistaken on this. My apologies as I didn't know that aspect of the aria-label requirements. If this is considered in the future, perhaps the link text can be automatically visible (not editable) in an aria-label field or preview area, but the aria-label attribute would only be inserted in the markup if the user types something in the aria-label field. Food for thought. Thank you for your time.
